src/buildercore/lifecycle.py

```diff
@@ -10,11 +10,37 @@
 import boto # route53 boto2 > route53 boto3
 from . import config, core
 from .core import boto_conn, find_ec2_instances, find_rds_instances, stack_all_ec2_nodes, current_ec2_node_id, NoPublicIps, NoRunningInstances
-from .utils import call_while, ensure
+from .utils import call_while, ensure, lmap
 from .context_handler import load_context, download_from_s3
 
 LOG = logging.getLogger(__name__)
 
+def _node_id(node):
+    name = core.tags2dict(node.tags)['Name']
+    return core.parse_stackname(name, all_bits=True, idx=True).get('cluster_id', 1)
+
+def ec2_nodes(stackname):
+    "returns all non-terminated nodes associated with stackname, ordered by node-id"
+    nodes = find_ec2_instances(stackname, state='pending|running|shutting-down|stopping|stopped')
+    return sorted(nodes, key=_node_id)
+
+def start_rds_nodes(stackname):
+    rds_to_be_started = _rds_nodes_states(stackname)
+    LOG.info("RDS nodes to be started: %s", rds_to_be_started)
+
+    def _start(nid):
+        return _rds_connection(stackname).start_db_instance(DBInstanceIdentifier=nid)
+    return lmap(_start, rds_to_be_started.keys())
+
+def restart(stackname):
+    "for each ec2 node in given stack, ensure ec2 node is stopped, then start it, then repeat. rds is started if stopped but otherwise not affected"
+    start_rds_nodes(core.find_rds_nodes(stackname))
+    for node in ec2_nodes(stackname):
+        node.stop()
+        node.wait_until_stopped()
+        node.start()
+        node.wait_until_running()
```

Comment:
DNS may need to be updated (although if there is more than one node, nothing needs to be done), see https://github.com/elifesciences/builder/pull/301/files#diff-e26cdd0bfc8a5ff0ffb7859c59362a18L195

Reply:
Also a running state is usually no guarantee that services have started, see https://github.com/elifesciences/builder/pull/301/files#diff-e26cdd0bfc8a5ff0ffb7859c59362a18R196 that waits for a standard file that signals the boot has been completed